plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,375 @@
+{-# LANGUAGE DeriveAnyClass    #-}
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PackageImports    #-}
+{-# LANGUAGE RecordWildCards   #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+module Ide.Plugin.Hlint
+  (
+    descriptor
+  --, provider
+  ) where
+import Refact.Apply
+import Control.Arrow ((&&&))
+import Control.DeepSeq
+import Control.Exception
+import Control.Lens ((^.))
+import Control.Monad
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Except
+import Data.Aeson.Types (ToJSON(..), FromJSON(..), Value(..))
+import Data.Binary
+import Data.Hashable
+import qualified Data.HashMap.Strict as Map
+import Data.Maybe
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import Data.Typeable
+import Development.IDE.Core.FileStore
+import Development.IDE.Core.OfInterest
+import Development.IDE.Core.Rules
+import Development.IDE.Core.Shake hiding (withIndefiniteProgress)
+import Development.IDE.Types.Diagnostics as D
+import Development.IDE.Types.Location
+import Development.Shake
+
+#ifdef GHC_LIB
+import Development.IDE.Core.RuleTypes (GhcSession(..))
+import Development.IDE.GHC.Util (hscEnv)
+import "ghc-lib" GHC hiding (DynFlags(..))
+import "ghc" GHC as RealGHC (DynFlags(..))
+import "ghc" HscTypes as RealGHC.HscTypes (hsc_dflags)
+import qualified "ghc" EnumSet as EnumSet
+import Language.Haskell.GhclibParserEx.GHC.Driver.Session as GhclibParserEx (readExtension)
+#else
+import GHC hiding (DynFlags(..))
+#endif
+
+import Ide.Logger
+import Ide.Types
+import Ide.Plugin
+import Ide.Plugin.Config
+import Ide.PluginUtils
+import Language.Haskell.HLint as Hlint
+import Language.Haskell.LSP.Core
+import Language.Haskell.LSP.Types
+import qualified Language.Haskell.LSP.Types      as LSP
+import qualified Language.Haskell.LSP.Types.Lens as LSP
+import System.FilePath (takeFileName)
+import System.IO (hPutStr, noNewlineTranslation, hSetNewlineMode, utf8, hSetEncoding, IOMode(WriteMode), withFile, hClose)
+import System.IO.Temp
+import Text.Regex.TDFA.Text()
+import GHC.Generics (Generic)
+
+-- ---------------------------------------------------------------------
+
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId = (defaultPluginDescriptor plId)
+  { pluginRules = rules
+  , pluginCommands =
+      [ PluginCommand "applyOne" "Apply a single hint" applyOneCmd
+      , PluginCommand "applyAll" "Apply all hints to the file" applyAllCmd
+      ]
+    , pluginCodeActionProvider = Just codeActionProvider
+  }
+
+data GetHlintDiagnostics = GetHlintDiagnostics
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable GetHlintDiagnostics
+instance NFData   GetHlintDiagnostics
+instance Binary   GetHlintDiagnostics
+
+type instance RuleResult GetHlintDiagnostics = ()
```

Comment:
This is an unusual `RuleResult`. Perhaps add a comment explaining how this works?

Reply:
would [the explanation above](https://github.com/haskell/haskell-language-server/pull/166#discussion_r511645626) be enough? not sure why it is unusual, is it cause the rest of plugins are not generating diagnostics?

Reply:
no, it's because the result is `()`, so if I understand correctly this rule exists purely for the generation of diagnostics.

Reply:
Ok, added a comment about although i am not sure if it is useful enough, feel free to suggest another content